styleguide/angular.md

```diff
@@ -0,0 +1,178 @@
+
+Bindings in components are defined analogous to the previous scopes in directives:
+
+```Javascript 
+{
+    /**
+     * All bindings can be optional by adding a question mark after the operator:
+     *   oneWayBinding: '<?'
+     * All bindings can be aliased by adding a name after the operator:
+     *   oneWayBinding: '<aOneWayBoundProperty'
+     */
+    bindings: {
+		oneWayBinding: '<',
+		twoWayBinding: '=',
+        expressionBinding: '&',
+        aString: '@',
+    }
+}
+```
+
+1. **One-Way-Binding:** Use them for values you don't want to change bottom-up. Even though you can propagate changes (they are passed by reference!) don't do it.
+2. **Two-Way-Binding:** Use them for values that need to be synchronised between components. Use them with caution since they will heavily affect  performance of the application!
+3. **Expression-Binding:** Will wrap the expression in a callback and you can call it in your component's controller. This is especially useful for passing callbacks. Be aware that you have to call the function with an object that passes its properties as parameters to the function (tracked by the name of the parameter): 
+```HTML
+<my-component expression-binding="$ctrl.selectValue(result)"></my-component>
+```
+```Javascript 
+class MyComponentController {
+   onValueSelected(selected) {
+      // property of the object is called result as specified in the expression
+      this.expressionBinding({ result: selected });
+   }
+}
+```
+4. **String-Binding:** interpolated string value
+
+#### One-Time-Binding
+
+A special - but important - case are [one-time bindings](https://docs.angularjs.org/guide/expression#one-time-binding), a an expression syntax used in your views. It will not trigger re-rendering values as soon as the expression has stabilised. **Use it for values that will not change anymore once they are present. This will increase the performance of your Angular app, especially if used in loops.**
+
+Let's add an example here. For SBC we have some sort of language panels for the translation of entities. Each panel renders a form and can be accessed using a tab. Therefore we iterate over the languages we load from the api. These languages will not change in real time, so we use a one time binding (note the double-colon in front of the expression):
+
+```HTML 
+<ul>
+    <li ng-repeat="language in ::$ctrl.languages">
+       <translation-form language="language">
+       </translation-form>
+    </li>
+</ul>
+```
+
+#### Ng-if vs. Ng-show
+
+Do not hide content using `ng-show` as long as you really need the content to be processed! Be aware that `ng-show` will hide an element in the DOM but it will be processed by Angular and and impact your performance.
+
+### Testing
+
+> **Discussion:** We do not have a proper testing strategy.  
+
+## General
+Discussion of general problems and approaches.
+
+### Global Variables
```

Comment:
how about a central vanilla js service that manages values that need to be passed from the back- to the frontend? This would give us a central standardised way of communication. The example may be way off since i don't know a thing about how the communication works at the moment.

this code is injected at the top of the page
```javascript
class GobalValueStorage {
    constructor() {
         this.values = new Map();
    }

    set(key, value) {
        if (this.values.has(key)) throw new Error(`GobalValueStorage: cannot set value ${key}, it was already set before`);
        this.values.set(key, value);
    }

    get(key) {
         return this.values.get(key);
   }

    has(key) {
         return this.values.has(key);
   }
}
window.sbcGlobals = new GobalValueStorage();
```

this is injected by typo3
```javascript
window.sbcGlobals.set('x', 4);
```

Reply:
Fair point. Let's discuss this. We'll have to check for limitations of the CMS in terms of JS code injection.

Reply:
@eventEmitter just noticed that we could never add values to your storage ;)